pandas/core/generic.py

```diff
@@ -728,7 +730,7 @@ def _set_axis(self, axis, labels):
         self._data.set_axis(axis, labels)
         self._clear_item_cache()
 
-    def transpose(self, *args, **kwargs):
+    def transpose(self, *args, **kwargs) -> "NDFrame":
```

Comment:
If we start adding type hints to public functions (maybe we already have other places where we do this, but just noticing it here): we should think about what to do with "NDFrame". 
This is not a "public term", and IMO no user should see this.

Should it be replaced with a union of Series and Frame?

Reply:
@simonjayhawkins should this by `pd._typing._T`?

Reply:
> @simonjayhawkins should this by `pd._typing._T`?

`pd._typing._T` is for unbound types (normally objects in a generic container). we should use `TypeVar("FrameOrSeries", bound="NDFrame")` in `core.generic`

this is currently `pd._typing.FrameOrSeries`. some potentially relevant discussion in #29480 

> This is not a "public term", and IMO no user should see this.

that's a reasonable agrument and perhaps also relevant to #29480